### ISSUE TITLE
`possibly-undefined` false positive on `for...else` loop

### ISSUE THREAD

**Issue by aardjon** [python/mypy#19985]
**Bug Report**

Mypy falsely reports `possibly-undefined` for a variable set within a `for`...`else` search.

**To Reproduce**

```python
def prick(straw: str) -> int: ...

def find_needle(haystack: Collection[str], needle: str) -> int:
    for straw in haystack:
        if straw == needle:
            pain = prick(needle)
            break
    else:
        raise ValueError("No needle found in this haystack.")

    return pain  # Mypy: Name "pain" may be undefined  [possibly-undefined]
```

**Expected Behavior**

I expect mypy to not complain on the `find_needle` function, because:
 - If the loops finds a match, the variable is set
 - If the loops ends without a match, the `else` branch is executed and the function raises
 - If `haystack` is empty, the `else` branch is executed and the function raises

(If I am wrong, please correct me!)

**Actual Behavior**

Mypy complains about `Name "pain" may be undefined` on the last line, though I really don't see a path in which the `pain` variable can be undefined there.

**Your Environment**

- Mypy version used: 1.18.2
- Mypy command-line flags: None
- Mypy configuration options from `pyproject.toml`: `strict = true`, `enable_error_code = ["possibly-undefined",]`
- Python version used: 3.12.3
